aderyn_core/src/detect/high/msg_value_in_loops.rs

```diff
@@ -0,0 +1,140 @@
+use std::collections::BTreeMap;
+use std::convert::identity;
+use std::error::Error;
+
+use crate::ast::{ASTNode, Expression, NodeID};
+
+use crate::capture;
+use crate::context::browser::ExtractMemberAccesses;
+use crate::context::investigator::{
+    StandardInvestigationStyle, StandardInvestigator, StandardInvestigatorVisitor,
+};
+use crate::detect::detector::IssueDetectorNamePool;
+use crate::{
+    context::workspace_context::WorkspaceContext,
+    detect::detector::{IssueDetector, IssueSeverity},
+};
+use eyre::Result;
+
+#[derive(Default)]
+pub struct MsgValueUsedInLoopDetector {
+    // Keys are: [0] source file name, [1] line number, [2] character location of node.
+    // Do not add items manually, use `capture!` to add nodes to this BTreeMap.
+    found_instances: BTreeMap<(String, usize, String), NodeID>,
+}
+
+impl IssueDetector for MsgValueUsedInLoopDetector {
+    fn detect(&mut self, context: &WorkspaceContext) -> Result<bool, Box<dyn Error>> {
+        // Investigate for loops to check for usage of `msg.value`
+        for for_statement in context.for_statements() {
+            if uses_msg_value(context, &(for_statement.into())).is_some_and(identity) {
+                capture!(self, context, for_statement);
+            }
+        }
+
+        // Investigate while loops to check for usage of `msg.value`
+        for while_statement in context.while_statements() {
+            if uses_msg_value(context, &(while_statement.into())).is_some_and(identity) {
+                capture!(self, context, while_statement);
+            }
+        }
+
+        // Investigate the do while loops to check for usage of `msg.value`
+        for do_while_statement in context.do_while_statements() {
+            if uses_msg_value(context, &(do_while_statement.into())).is_some_and(identity) {
+                capture!(self, context, do_while_statement);
+            }
+        }
+
+        Ok(!self.found_instances.is_empty())
+    }
+
+    fn severity(&self) -> IssueSeverity {
+        IssueSeverity::High
+    }
+
+    fn title(&self) -> String {
+        String::from("Loop contains `msg.value`.")
+    }
+
+    fn description(&self) -> String {
+        String::from("Do not use `msg.value` in loops since it may be misleading when calculating the total amount.\
+            It's recommended to prepare additional variables outside the loop to do the required calculation.")
```

Comment:
Let's be more explicit here. Taken from the Slither wiki:

"Provide an explicit array of amounts alongside the receivers array, and check that the sum of all amounts matches msg.value."

Reply:
Hmm this one I thought was too specific of a description. But I think explicit is good. We'll do that :)